src/index.js

```diff
@@ -50,6 +65,155 @@ function getTriggeringProps<PropsInput: {}>(
 
 const prefetchTimeout = 2000 // ms
 
+// TODO: This is probably not going to be 100% safe to use under React async mode
+// Do more research
+class WithObservablesComponent<AddedValues: any, PropsInput: {}> extends Component<
+  *,
+  {
+    isFetching: boolean,
+    values: $FlowFixMe<AddedValues>,
+    triggeredFromProps: any[],
+  },
+> {
+  BaseComponent: React$ComponentType<Object>
+
+  triggerProps: TriggerProps<PropsInput>
+
+  getNewProps: PropsInput => Observable<Object>
+
+  _subscription: ?Subscription = null
+
+  _isMounted = false
+
+  _prefetchTimeout: ?TimeoutID = null
+
+  _exitedConstructor = false
+
+  constructor(
+    props,
+    BaseComponent: React$ComponentType<Object>,
+    getNewProps: PropsInput => Observable<Object>,
+    triggerProps: TriggerProps<PropsInput>,
+  ): void {
+    super(props)
+    this.BaseComponent = BaseComponent
+    this.triggerProps = triggerProps
+    this.getNewProps = getNewProps
+    this.state = {
+      isFetching: true,
+      values: {},
+      triggeredFromProps: getTriggeringProps(props, triggerProps),
+    }
+
+    // The recommended React practice is to subscribe to async sources on `didMount`
+    // Unfortunately, that's slow, because we have an unnecessary empty render even if we
```

Comment:
Hmm I am thinking actually how slow is this empty render and subscribing on mount? It adds a lot of complexity and as you stated might not by compatible with future React changes.

Reply:
This part should work with Concurrent Mode - mobx-lite does something similar. the UNSAFE_xxx handler is unsafe for concurrent mode.

how slow -- it's hard for me to say exactly right now, but the problem is that did mount is called after the changes are committed. So it has to go through DOM/native code before the second pass can occur -- even if changes are available synchronously. I guess that's the main bottleneck.

I will do some research on this issue